statsd/aggregator.go

```diff
@@ -341,48 +338,45 @@ func (a *MetricAggregator) receiveMetric(m *types.Metric, now time.Time) {
 		log.Errorf("Unknow metric type %s for %s", m.Type, m.Name)
 	}
 
-	a.Stats.LastMessage = time.Now()
+	a.Stats.LastMessage = now
 }
 
-func (a *MetricAggregator) processQueue() {
-	for metric := range a.MetricQueue {
-		a.receiveMetric(metric, time.Now())
-		runtime.Gosched()
-	}
-}
-
-// Aggregate starts the MetricAggregator so it begins consuming metrics from MetricChan
+// Aggregate starts the MetricAggregator so it begins consuming metrics from MetricQueue
 // and flushing them periodically via its Sender.
 func (a *MetricAggregator) Aggregate() {
-	flushChan := make(chan error)
 	flushTimer := time.NewTimer(a.FlushInterval)
 
-	for i := 0; i < a.MaxWorkers; i++ {
-		go a.processQueue()
-	}
-
 	for {
 		select {
+		case metric, ok := <-a.MetricQueue:
+			if ok {
+				a.receiveMetric(metric, time.Now())
```

Comment:
Only this goroutine now handles metrics from the queue. To handle a metric we need to lock the mutex - no point in trying to do this concurrently.